src/nupic/regions/SPRegion.cpp

```diff
@@ -0,0 +1,945 @@
+  NTA_CHECK(outputBuffer.getType() == NTA_BasicType_SDR);
+  UInt32 columnCount = (UInt32)outputBuffer.getCount();
+  if (columnCount == 0 || outputBuffer.getBuffer() == nullptr) {
+    NTA_THROW << "SPRegion::initialize - Output buffer not set.\n";
+  }
+
+  // All input links should have been initialized during Network.initialize().
+  // However, if nothing is connected it might not be. The SpatialPooler
+  // algorithm requires input.
+  //
+  // If there are more than one input link (FAN-IN), the input buffer will be the
+  // concatination of all incomming buffers.  
+  Input *in = getInput("bottomUpIn");
+  if (!in->hasIncomingLinks())
+     NTA_THROW << "SPRegion::initialize - No input links were configured for this SP region.\n";
+  Array &inputBuffer = in->getData();
+  NTA_CHECK(inputBuffer.getType() == NTA_BasicType_SDR);
+  args_.inputWidth = (UInt32)inputBuffer.getCount();
+  if (args_.inputWidth == 0) {
+    NTA_THROW << "SPRegion::initialize - No input buffer was allocated for this SP region.\n";
+  }
+
+  // Take the dimensions directly from the SDRs.
+  std::vector<UInt32> inputDimensions = inputBuffer.getSDR()->dimensions;
+  std::vector<UInt32> columnDimensions = outputBuffer.getSDR()->dimensions;
+  if (args_.potentialRadius == 0)
+    args_.potentialRadius = args_.inputWidth;
+
+  // instantiate a SpatialPooler.
+  sp_ = std::unique_ptr<SpatialPooler>(new SpatialPooler(
+      inputDimensions, columnDimensions, args_.potentialRadius,
+      args_.potentialPct, args_.globalInhibition, args_.localAreaDensity,
+      args_.numActiveColumnsPerInhArea, args_.stimulusThreshold,
+      args_.synPermInactiveDec, args_.synPermActiveInc, args_.synPermConnected,
+      args_.minPctOverlapDutyCycles, args_.dutyCyclePeriod, args_.boostStrength,
+      args_.seed, args_.spVerbosity, args_.wrapAround));
+}
+
+
+
+void SPRegion::compute() {
+  NTA_ASSERT(sp_) << "SP not initialized";
+
+  // BOTTOM-UP compute mode
+  args_.iter++;
+
+  // prepare the input
+  Array &inputBuffer  = getInput("bottomUpIn")->getData();
+  Array &outputBuffer = getOutput("bottomUpOut")->getData();
+
+  // Call SpatialPooler compute
+  sp_->compute(*inputBuffer.getSDR(), args_.learningMode, *outputBuffer.getSDR());
+
+}
+
+std::string SPRegion::executeCommand(const std::vector<std::string> &args,Int64 index) {
+  // The Spatial Pooler does not execute any Commands.
+  return "";
+}
+
+// This is the per-node output size. This is called by Link to determine how big 
+// to create the output buffers during Region::initialization(). It calls this
+// only if dimensions were not set on this region.
+// NOTE: Some outputs are optional, return 0 if not used.
+size_t SPRegion::getNodeOutputElementCount(const std::string &outputName) {
+  if (outputName == "bottomUpOut") // This is the only output link we actually use.
+  {
+      return args_.columnCount; 
+  }
+  return 0; // an optional output that we don't use.
+}
+
+
+
+Spec *SPRegion::createSpec() {
```

Comment:
maybe move this method to hpp? 
It's lot of lines, mostly doc, so it would fit there. (for your consideration only)